overlord/snapstate/handlers.go

```diff
@@ -4006,45 +4007,55 @@ func refreshedSnaps(reTask *state.Task) (snapNames []string, failed bool) {
 		if task.Kind() == "check-rerefresh" {
 			// we've reached a previous check-rerefresh (but not ourselves).
 			// Only snaps in tasks after this point are of interest.
-			laneSnaps = map[int]string{}
-		}
-		lanes := task.Lanes()
-		// Tasks can have multiple lanes, but can also share a single lane, right now we rely on
-		// the first lane, which may not be sufficient to detect whether or not a specific
-		// snap has been refreshed.
-		// XXX: If snaps have been refreshed with Flags.Transaction == "all-snap", then this may
-		// be an issue and not accurately detect which snaps have been refreshed.
-		lane := lanes[0]
-		if lane == 0 {
-			// not really a lane
-			continue
+			laneSnaps = make(map[int]map[string]bool)
 		}
-		if task.Status() != state.DoneStatus {
-			// ignore non-successful lane (1)
-			laneSnaps[lane] = ""
+
+		// Ignore tasks on '0' lane, they are not refreshes anyway.
+		taskLanes := task.Lanes()
+		if len(taskLanes) == 1 && taskLanes[0] == 0 {
 			continue
 		}
-		if _, ok := laneSnaps[lane]; ok {
-			// ignore lanes we've already seen (including ones explicitly ignored in (1))
-			continue
+
+		// Track lanes that failed.
+		if task.Status() != state.DoneStatus {
+			for _, l := range taskLanes {
+				failedLanes[l] = true
+			}
 		}
+
+		// Only check "prepare-snap" and "download-snap" as they are the tasks that have
+		// "snap-setup" attached to them. No point in checking every task in task-sets.
 		var snapsup SnapSetup
-		if err := task.Get("snap-setup", &snapsup); err != nil {
+		switch task.Kind() {
+		case "prepare-snap", "download-snap":
```

Comment:
nitpick: as I wrote in my original comment we probably can consider just download-snap, as prepare-snap is only used for local only operations which means we did not interrogate the store there so unclear why we would do it now, as the whole helper is meant for re-refresh only

Reply:
Ah right, changed this and updated unit tests that were using "prepare-snap".